### PR TITLE
Wmwragg/direct chat sublist

### PR DESCRIPTION
**Pull request by wmwragg**
A companion branch vector-im/vector-web:wmwragg/direct-chat-sublist also needs to be merged.

Signed-off-by: William Wragg wm.wragg@gmail.com

### CODE REVIEW

File: src/MatrixTools.js [hunk 1]
```diff
@@ -24,5 +25,34 @@ module.exports = {
     getDisplayAliasForRoom: function(room) {
         return room.getCanonicalAlias() || room.getAliases()[0];
     },
+
+    isDirectMessageRoom: function(room, me, ConferenceHandler, hideConferenceChans) {
```
Comment (ara4n) [inline src/MatrixTools.js]: Heads up that this function will change shortly once we've finally finally defined how to track which rooms are DMs in terms of room or user configuration state.  This is fine for now though.

File: src/components/views/rooms/RoomList.js [hunk 2]
```diff
@@ -63,6 +64,13 @@ module.exports = React.createClass({
 
     componentDidMount: function() {
         this.dispatcherRef = dis.register(this.onAction);
+        // Initilaise the stickyHeaders when the component is created
```
Comment (ara4n) [inline src/components/views/rooms/RoomList.js]: s/Initilaise/Initialise/

File: src/components/views/rooms/RoomList.js [hunk 3]
```diff
@@ -63,6 +64,13 @@ module.exports = React.createClass({
...
+        this._updateStickyHeaders(true);
+    },
+
+    componentDidUpdate: function() {
+        // Reinitilaise the stickyHeaders when the component is updated
```
Comment (ara4n) [inline src/components/views/rooms/RoomList.js]: s/Reinitilaise/Reinitialise/

File: src/components/views/rooms/RoomList.js [hunk 4]
```diff
@@ -317,6 +322,99 @@ module.exports = React.createClass({
         }
     },
 
+    // Doing the stckiy headers as raw DOM, for speed, as it gets very stuttery if done
```
Comment (ara4n) [inline src/components/views/rooms/RoomList.js]: s/stckiy/sticky/

File: src/components/views/rooms/RoomList.js [hunk 5]
```diff
@@ -317,6 +322,99 @@ module.exports = React.createClass({
...
+        // Use the offset of the top of the componet from the window
+        // as this is used to calculate the CSS fixed top position for the stickies
+        var scrollAreaHeight = ReactDOM.findDOMNode(this).getBoundingClientRect().height;
+
+        if (initialise) {
+            // Get a collection of sticky header containers references
+            this.stickies = document.getElementsByClassName("mx_RoomSubList_labelContainer");
+
+            // Make sure there is sufficient space to do sticky headers
+            this.scrollAreaSufficient = (120 + (this.stickies[0].getBoundingClientRect().height * this.stickies.length)) < scrollAreaHeight;
```
Comment (ara4n) [inline src/components/views/rooms/RoomList.js]: perhaps give 120 a variable name or comment to say where it comes from?

**Comment by ara4n** [comment]
LGTM other than minor comments.

File: src/MatrixTools.js (hunk 1 above)
Comment (wmwragg) [inline src/MatrixTools.js]: Yep, made it a separate function, as you mentioned this when I was doing it, so this should be easily swapped out for the new behaviour once it's implemented.

**Comment by ara4n** [comment]
@matrixbot okay to test

File: src/components/views/rooms/RoomList.js (hunk 2 above)
Comment (wmwragg) [inline src/components/views/rooms/RoomList.js]: done

File: src/components/views/rooms/RoomList.js (hunk 3 above)
Comment (wmwragg) [inline src/components/views/rooms/RoomList.js]: done

File: src/components/views/rooms/RoomList.js (hunk 4 above)
Comment (wmwragg) [inline src/components/views/rooms/RoomList.js]: done

File: src/components/views/rooms/RoomList.js (hunk 5 above)
Comment (wmwragg) [inline src/components/views/rooms/RoomList.js]: Added further explanation that sticky headers requires 120 pixels plus all the sticky headers

**Pull request merged**